daemon: refactor translateContainerdStartErr() and remove unused argument

thaJeztah opened the pull request:
### daemon: translateContainerdStartErr() remove unused cmd argument

This argument was no longer used since commit https://github.com/moby/moby/commit/225e046d9d1bdf0f06f4c97bef239b5909463885 (part of https://github.com/moby/moby/pull/42270)

for further details; look at individual commit messages

thaJeztah committed:
daemon: translateContainerdStartErr(): remove unused cmd argument

This argument was no longer used since commit 225e046d9d1bdf0f06f4c97bef239b5909463885

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on daemon/exec.go:
```diff
@@ -267,10 +267,13 @@ func (daemon *Daemon) ContainerExecStart(ctx context.Context, name string, stdin
+		if exitCode != -1 {
+			ec.SetExitCode(exitCode)
+		}
 		ec.Unlock()
```
This moves the `Unlock()` to happen _after_ the exit-code was updated.

Possibly the `c.ExecCommands.Unlock()` should be moved as well (not sure if it matters when, and in which order they're unlocked)

thaJeztah commented on daemon/errors.go:
```diff
@@ -132,35 +132,46 @@ func (e startInvalidConfigError) Error() string {
+// container and an errdefs error (either errdefs.ErrInvalidParameter or errdefs.ErrUnknown).
+//
+// exitCode can be one of:
+// - -1:   unknown error (do not update the container's exitcode)
```
FWIW; I was a bit in doubt what the most logical value would be for "we don't have a new exit code for you"; we could use `0` (default for `int`), but possibly this would be confusing (as `0` means "no error"); So, I landed on `-1`, which is not a valid exit-code, and as indicator "we don't know, so keep it as-is"

Perhaps it could be changed to `1` (there was an error!) But this would potentially be overwriting the actual container's exit code? (not sure if it's set already by this time though)

thaJeztah commented:
Arf; looks like I didn't stage/push one change. Will fix when I'm back at my keyboard

tianon commented on daemon/start.go:
```diff
@@ -189,21 +189,26 @@ func (daemon *Daemon) containerStart(container *container.Container, checkpoint
-		if err != nil {
-			return translateContainerdStartErr(container.SetExitCode, err)
+		if exitCode, err := translateContainerdStartErr(err); err != nil {
+			if exitCode != -1 {
```
Maybe we should make a constant for this "magic" number?

thaJeztah replied:
Heh; so I initially creating `const's` for the possible exit-codes, but the descriptions became a bit "fuzzy" (especially the `127` being "could be this, or could be that"), so I removed them, but yeah, perhaps I should have some consts (it's not exported, so I guess they don't have to be squeaky clean on description.

tianon replied:
Yeah, I think this is the only one that stuck out because it's the only one I see special parsing for, likely because it's the only one that's new? :innocent:

thaJeztah replied:
I pushed a commit with const/enums (but I feel like it's a bit too much perhaps). Alternatively, I could change the checks to `if exitCode > 0` do you think that would make it clearer?

thaJeztah replied:
Actually, perhaps I should just make sure that it always returns a sensible exit-code. I wasn't sure if the exitCode could be set through other means, but looking at both `state.SetExitCode()`, and `state.ExitCodeValue`, then it looks like all updates happen within this file:

<img width="898" alt="Screenshot 2022-02-16 at 20 12 33" src="https://user-images.githubusercontent.com/1804568/154339645-2b42fa6f-79f3-4c23-bff7-611a33d7fc57.png">

<img width="595" alt="Screenshot 2022-02-16 at 20 17 29" src="https://user-images.githubusercontent.com/1804568/154339650-a27d57b2-a7f0-4d3b-93fd-9e1709ca487c.png">


And if it's not set, it's updated to `128` in the `defer()` https://github.com/moby/moby/blob/9674540ccff358c3cd84cc2f33c3503e0dab7fb7/daemon/start.go#L124-L127

From that, I'm tempted to say that we can return `128` as a default (for "unknown error")

thaJeztah replied:
Arf.. I may be looking at the wrong `SetExitcode()` (state vs config?); let me get back on you on that

thaJeztah replied:
Pushed a commit with that change; I can squash some of the commits

thaJeztah commented on daemon/start_unix.go:
```diff
@@ -16,8 +16,11 @@ func (daemon *Daemon) getLibcontainerdCreateOptions(container *container.Contain
 	rt, err := daemon.getRuntime(container.HostConfig.Runtime)
-	if err != nil {
-		return "", nil, translateContainerdStartErr(container.Path, container.SetExitCode, err)
+	if exitCode, err := translateContainerdStartErr(err); err != nil {
+		if exitCode != -1 {
+			container.SetExitCode(exitCode)
+		}
+		return "", nil, err
```
This was the bit I didn't stage, but looking at this again, I don't think this handling makes sense at all here;

`translateContainerdStartErr()` is used to first convert a grpc-error coming from containerd; https://github.com/moby/moby/blob/72fef53cecdd0b151ecaa8301b6d26f5d3182e9f/daemon/errors.go#L136

Which checks if it's a GRPC error, and if not "creates a new error" with the original error message string; https://github.com/moby/moby/blob/72fef53cecdd0b151ecaa8301b6d26f5d3182e9f/vendor/google.golang.org/grpc/status/status.go#L148-L153

However, the error here comes from `daemin.getruntime()`, which; already can return a typed error (if the configured runtime is incorrect);

https://github.com/moby/moby/blob/72fef53cecdd0b151ecaa8301b6d26f5d3182e9f/daemon/runtime_unix.go#L134
https://github.com/moby/moby/blob/72fef53cecdd0b151ecaa8301b6d26f5d3182e9f/daemon/runtime_unix.go#L152

None of which are related to the container's `cmd` nor with (bind) mounts.

Then, there's a third possible error, coming from `daemon.rewriteRuntimePath()` https://github.com/moby/moby/blob/72fef53cecdd0b151ecaa8301b6d26f5d3182e9f/daemon/runtime_unix.go#L138-L140

The only error that can be returned there, is if `exec.Lookpath()` fails to find the runtime; https://github.com/moby/moby/blob/72fef53cecdd0b151ecaa8301b6d26f5d3182e9f/daemon/runtime_unix.go#L124-L126

I'm tempted to return those errors as-is (do we need to update the container exit status? perhaps). _However_, this is all on container _start_ (not create), so what's the appropriate response? There's nothing wrong with the request itself ("start this container"), but the configuration of the container (on disk) may be incorrect, so a `500` (internal server error) would technically be more appropriate.

The bad thing is that we perform this validation on container _start_ (not on container _create_), which would possibly be appropriate (probably _in addition_ to the current check, as the runtime may _become_ invalid, so should cause an error when restarting the daemon (and container) in that case).

thaJeztah committed:
daemon: translateContainerdStartErr(): use const/enum for exit-statuses

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

tianon commented on daemon/errors.go:
```diff
@@ -132,35 +132,50 @@ func (e startInvalidConfigError) Error() string {
+// container and an errdefs error (either errdefs.ErrInvalidParameter or errdefs.ErrUnknown).
+func translateContainerdStartErr(err error) (exitCode exitStatus, exitErr error) {
+	if err == nil {
+		return exitNoError, nil
```
Hmm shouldn't this case _also_ leave the container's exit status alone?  (Sorry for not being more familiar with this code :see_no_evil:)

Looking at the previous implementation, it seems to not do anything very different in the `nil` case - maybe just because the function wasn't called in the `nil` case?  (The whole translation feels odd IMO, but I get it.)

thaJeztah replied:
>  maybe just because the function wasn't called in the nil case?

Yes, so previously the container state was modified within this function, and the function wasn't called if `err`  was nil

IN this PR the update happens if `err != nil`, so basically "any value" would be ok here;

```go
if exitCode, err := translateContainerdStartErr(err); err != nil {
	container.SetExitCode(exitCode)
	return err
}
```

(So if `err` is nil to begin with, it would not update the exit code)

tianon replied:
Ahhh got it, thanks for clarifying :+1:

thaJeztah committed:
daemon: translateContainerdStartErr(): use exit code 128 as default

After further looking at the code, it appears that the default exit-code
for unknown (other) errors is 128 (set in `defer`).

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
daemon: translateContainerdStartErr(): extract detecting wrong cmd

To make the code slightly more readable, and slightly DRY.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
daemon: translateContainerdStartErr(): use early return for errors

These matches were overwriting the previous "match", so reversing the
order in which they're tried so that we can return early.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
daemon: translateContainerdStartErr(): rename to setExitCodeFromError()

This should hopefully make it slightly clearer what it does.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
@tianon I cleaned up this PR, and reverted the "signature" change of the function, instead opted to rename it to something (hopefully) more obvious. PTAL (individual commits describe what I did)

thaJeztah committed:
daemon: containerStart(): rename return variable

Rename the variable make it more visible where it's used, as there's were
other "err" variables masking it.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Rebased to get a fresh run of CI: ptal 🤗
